# bios8366_fp_pw_111418.ipynb

# 11.14.18

In [1]:
%matplotlib inline
import numpy as np
import pymc3 as pm
import pylab as plt
import pandas as pd
import seaborn as sns
sns.set(style="whitegrid")

# Set seed
np.random.seed(10011)

/anaconda2/envs/bios8366/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/anaconda2/envs/bios8366/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)


## LABS

1) create function looking at top X number of labs

In [2]:
# import labs
df_LAB = pd.read_csv('./Data/FONNESBECK_LAB_20151202.csv',sep=',',engine='python',quoting=3)
df_LAB.head()

RUID Lab_name    Lab_date Lab_value
0  50135262     %SAT  04/13/2008        17
1  50135262      ABO  02/08/2007         A
2  50135262      ABO  02/24/2008         a
3  50135262   AN-GAP  02/08/2007         9
4  50135262   AN-GAP  02/11/2011         8

In [3]:
df_LAB1 = df_LAB.copy()

In [4]:
# create function looking at top X number of labs
def top_labs(labs_df,num_labs=10):
    
    """function takes in lab in dataframe format and outputs 
    
        :param labs_df = lab dataframe with columns = {'RUID','Lab_name','Lab_date','Lab_value'}
        :param num_labs = top X of desired top labs to output from function, default = top 10
        
        :return = 
            top_num_labs: pandas dataframe with columns {'Lab_name','counts'} that contains only labs from top number of desired labs
            top_pt_labs: pandas dataframe with columns {'RUID', 'Lab_name', 'Lab_date','Lab_value'} that contains only labs from top number of desired labs
    """
    # preprocess lab dataframe to get sorted labs
    top_labs = pd.DataFrame(labs_df[['Lab_name']].groupby(['Lab_name']).size()).reset_index()
    top_labs.columns = ['Lab_name','counts']
    top_labs1 = top_labs.sort_values(by='counts',ascending=False)
    
    # get dataframe with columns = {'Lab_name', 'counts'}
    top_num_labs = top_labs1.iloc[0:int(num_labs),:]
    
    # create dataframe with only labs matching top labs
    top_pt_labs = labs_df[labs_df['Lab_name'].isin(top_num_labs['Lab_name'])]
    
    return top_num_labs, top_pt_labs

In [5]:
top_num_labs, top_pt_labs = top_labs(df_LAB1,10)

In [6]:
top_num_labs

Lab_name  counts
1762   GluBed  252113
2218        K  238492
1154    Creat  233493
2609       Na  230471
2776      PCV  229964
592       BUN  226556
1765     Gluc  226554
926       CO2  226256
1070       Cl  226059
2920   Plt-Ct  215131

In [7]:
top_pt_labs.head()

RUID Lab_name    Lab_date Lab_value
121  50135262      BUN  02/08/2007        17
122  50135262      BUN  02/11/2011        25
123  50135262      BUN  02/12/2011        23
124  50135262      BUN  02/13/2007        15
125  50135262      BUN  02/13/2011        17

In [8]:
# write top_pt_labs to csv to read in later
top_pt_labs.to_csv('top_pt_labs_111418.csv',index=False)

In [11]:
# test to see if output was successful
top_pt_labs1 = pd.read_csv('top_pt_labs_111418.csv')
top_pt_labs1.head()

RUID Lab_name    Lab_date Lab_value
0  50135262      BUN  02/08/2007        17
1  50135262      BUN  02/11/2011        25
2  50135262      BUN  02/12/2011        23
3  50135262      BUN  02/13/2007        15
4  50135262      BUN  02/13/2011        17

In [12]:
top_pt_labs['Lab_name'].unique()

array(['BUN', 'CO2', 'Cl', 'Creat', 'GluBed', 'Gluc', 'K', 'Na', 'PCV',
       'Plt-Ct'], dtype=object)

2) explore histograms and remove non-physiological values for labs

In [13]:
glubed_labs = top_pt_labs[top_pt_labs['Lab_name']=='GluBed']

In [14]:
glubed_labs.shape

(252113, 4)

In [15]:
glubed_labs.describe()

RUID
count  2.521130e+05
mean   5.362519e+07
std    6.056270e+05
min    5.013526e+07
25%    5.372966e+07
50%    5.373197e+07
75%    5.373448e+07
max    5.373642e+07

In [42]:
glubed_labs2 = glubed_labs.copy()

In [46]:
glubed_labs3 = glubed_labs2.sort_values(by='Lab_value',ascending=False)
glubed_labs3.head()

RUID Lab_name    Lab_date Lab_value
3045178  53730650   GluBed  10/23/2013   invalid
767913   53728380   GluBed  09/19/2010   Results
6967417  53734447   GluBed  12/15/2011   Results
1783639  53729342   GluBed  01/08/2010   Results
4613124  53732154   GluBed  12/12/2008   Results

there are bedside glucose values that aren't informative, e.g. "invalid", "results", "ordered", "Credited", ">600", etc.

In [47]:
# identify non-numeric values
glubed_labs3['Lab_value'] = pd.to_numeric(glubed_labs3['Lab_value'],errors='coerce')
glubed_labs3.head()

RUID Lab_name    Lab_date  Lab_value
3045178  53730650   GluBed  10/23/2013        NaN
767913   53728380   GluBed  09/19/2010        NaN
6967417  53734447   GluBed  12/15/2011        NaN
1783639  53729342   GluBed  01/08/2010        NaN
4613124  53732154   GluBed  12/12/2008        NaN

In [54]:
glubed_labs4 = glubed_labs3[glubed_labs3['Lab_value'].isna()]

In [55]:
glubed_labs4.shape

(1244, 4)

there are 1244 rows with NaN for `Lab_value`

In [56]:
# count number of unique patients with NaN
glubed_labs4['RUID'].nunique()

489

there 489 unique patients with NaN for `Lab_value`

In [57]:
# remove rows with NaN for Lab_value
glubed_labs5 = glubed_labs3[~glubed_labs3['Lab_value'].isna()]

In [61]:
glubed_labs6 = glubed_labs5.sort_values(by='Lab_value',ascending=False)
glubed_labs6.head()

RUID Lab_name    Lab_date  Lab_value
759604   53728363   GluBed  12/24/2013      600.0
1122181  53728649   GluBed  06/18/2014      593.0
2396782  53729950   GluBed  10/02/2013      593.0
8867857  53736205   GluBed  02/17/2014      593.0
8586755  53735988   GluBed  10/13/2014      593.0

In [62]:
# get summary stats for cleaned dataframe
glubed_labs6.describe()

RUID     Lab_value
count  2.508690e+05  250869.00000
mean   5.362500e+07     159.84525
std    6.061220e+05      74.08201
min    5.013526e+07      12.00000
25%    5.372966e+07     111.00000
50%    5.373197e+07     141.00000
75%    5.373448e+07     189.00000
max    5.373642e+07     600.00000

this cleaned up dataset looks better, in terms of summary stats

In [63]:
# export cleaned-up dataframe to csv
glubed_labs6.to_csv('glubed_labs6.csv',index=False)

In [65]:
# read in csv file as check
glubed_labs_import = pd.read_csv('glubed_labs6.csv')
glubed_labs_import.head()

RUID Lab_name    Lab_date  Lab_value
0  53728363   GluBed  12/24/2013      600.0
1  53728649   GluBed  06/18/2014      593.0
2  53729950   GluBed  10/02/2013      593.0
3  53736205   GluBed  02/17/2014      593.0
4  53735988   GluBed  10/13/2014      593.0

## BLOOD-PRESSURE

1) remove non-physiological measures of BP

In [66]:
# import BP data
df_BP = pd.read_csv('./Data/FONNESBECK_BP_20151202.csv',sep=',')
df_BP.head()

RUID  SYSTOLIC  DIASTOLIC Measure_date
0  50135262       150         80   01/09/2005
1  50135262       137         77   02/08/2007
2  50135262       137         78   02/08/2007
3  50135262       190         78   02/08/2007
4  50135262       190         77   02/08/2007

In [68]:
df_BP.shape

(10430396, 4)

In [74]:
df_BP['RUID'].nunique()

7484

In [71]:
# identify very high values, most likely inaccurate systolic blood pressures
df_BP1 = df_BP[df_BP['SYSTOLIC']>400]

In [73]:
# see how many unique patients in dataframe
df_BP1['RUID'].nunique()

225

there are 225/7484 patients in dataframe with very high systolic BP

In [83]:
pd.set_option('display.max_rows', 1000)
df_BP1.sort_values(by='SYSTOLIC',ascending=True).tail(1000)

RUID  SYSTOLIC  DIASTOLIC Measure_date
6771804   53733372       944         69   08/11/2014
7360064   53733995       944         75   07/01/2012
7360072   53733995       944         64   07/01/2012
7360073   53733995       944         44   07/01/2012
9425034   53735673       944         39   04/13/2007
7360077   53733995       944         60   07/01/2012
7360056   53733995       944         80   07/01/2012
9853531   53736042       944         57   04/18/2013
9972680   53736120       945         70   09/10/2005
3348248   53730650       945         47   06/01/2010
3348255   53730650       945         45   06/01/2010
3348272   53730650       945         58   06/01/2010
6875755   53733455       945         58   06/28/2006
2075558   53729383       945         65   07/04/2013
2891967   53730159       945         64   11/04/2010
2891970   53730159       945         63   11/04/2010
4025886   53731119       945          7   09/30/2011
1943123   53729264       945         72   03/05/2008
4025906   53731119       945         55   09/30/2011
4025891   53731119       945         61   09/30/2011
1943116   53729264       945          0   03/05/2008
4025892   53731119       945         41   09/30/2011
1943110   53729264       945         50   03/05/2008
1943073   53729264       945         58   03/05/2008
2075535   53729383       945         53   07/04/2013
2075540   53729383       945          7   07/04/2013
2075555   53729383       945         51   07/04/2013
4025903   53731119       945         57   09/30/2011
3348281   53730650       945         52   06/01/2010
2891951   53730159       945         69   11/04/2010
2075557   53729383       945         61   07/04/2013
2075563   53729383       945         62   07/04/2013
3348302   53730650       945         43   06/01/2010
3348385   53730650       945          2   06/01/2010
3348317   53730650       945         55   06/01/2010
8571424   53734978       945         55   02/19/2006
3348313   53730650       945         57   06/01/2010
2075564   53729383       945         82   07/04/2013
9972664   53736120       945         59   09/10/2005
9972670   53736120       945          9   09/10/2005
9972674   53736120       945         62   09/10/2005
4597292   53731614       945         66   03/26/2014
9972652   53736120       945         56   09/10/2005
6875754   53733455       945         69   06/28/2006
8571414   53734978       945         56   02/19/2006
8571403   53734978       945         60   02/19/2006
8571395   53734978       945         66   02/19/2006
8571392   53734978       945         58   02/19/2006
4597290   53731614       945         70   03/26/2014
4597281   53731614       945         63   03/26/2014
6489248   53733131       945         56   06/19/2008
3348432   53730650       945         59   06/01/2010
4844031   53731739       945         58   12/22/2013
8571422   53734978       945         88   02/19/2006
4844037   53731739       945         54   12/22/2013
3348332   53730650       945         46   06/01/2010
3348365   53730650       945         49   06/01/2010
3348377   53730650       945         42   06/01/2010
4597277   53731614       945         59   03/26/2014
3348415   53730650       945         53   06/01/2010
9972677   53736120       945         58   09/10/2005
4844032   53731739       945         49   12/22/2013
9972675   53736120       945         64   09/10/2005
1943095   53729264       945         67   03/05/2008
2891952   53730159       945          2   11/04/2010
4597269   53731614       945          4   03/26/2014
6875750   53733455       945         61   06/28/2006
6875751   53733455       945          1   06/28/2006
6489250   53733131       945          4   06/19/2008
6489249   53733131       945         50   06/19/2008
1943067   53729264       945         52   03/05/2008
2891964   53730159       945         67   11/04/2010
3348429   53730650       945         54   06/01/2010
1943053   53729264       945         54   03/05/2008
4906540   53731788       946         52

In [75]:
df_BP1.describe()

RUID     SYSTOLIC    DIASTOLIC
count  1.987000e+03  1987.000000  1987.000000
mean   5.373214e+07   908.807247    54.746855
std    2.537387e+03   111.616889    19.893494
min    5.372791e+07   411.000000     0.000000
25%    5.372984e+07   904.000000    47.000000
50%    5.373229e+07   944.000000    57.000000
75%    5.373442e+07   975.000000    66.500000
max    5.373631e+07   997.000000   155.000000

looking at the summary stats, it may make sense to drop these patients if we use blood pressure

2) another method is to make blood pressures over a certain range `NaN` and perhaps impute them

In [92]:
df_BP_temp = df_BP.copy()

In [95]:
df_BP_temp[df_BP_temp['RUID'].isna()]

Empty DataFrame
Columns: [RUID, SYSTOLIC, DIASTOLIC, Measure_date]
Index: []

In [99]:
pd.set_option('display.max_rows', 25)
df_BP_impute = df_BP.copy()
df_BP_impute

RUID  SYSTOLIC  DIASTOLIC Measure_date
0         50135262       150         80   01/09/2005
1         50135262       137         77   02/08/2007
2         50135262       137         78   02/08/2007
3         50135262       190         78   02/08/2007
4         50135262       190         77   02/08/2007
5         50135262       137         64   02/08/2007
6         50135262       186         78   02/08/2007
7         50135262       186         64   02/08/2007
8         50135262       190         64   02/08/2007
9         50135262       186         77   02/08/2007
10        50135262       160         72   02/09/2007
11        50135262       161         56   02/09/2007
...            ...       ...        ...          ...
10430384  53736422       114         65   05/05/2006
10430385  53736422       125         69   05/05/2006
10430386  53736422       114         75   05/05/2006
10430387  53736422       137         61   05/05/2006
10430388  53736422       122         66   05/05/2006
10430389  53736422       118         66   05/05/2006
10430390  53736422       118         73   05/05/2006
10430391  53736422       119         69   05/05/2006
10430392  53736422       121         59   05/05/2006
10430393  53736422       160         80   06/07/2006
10430394  53736422       138         77   06/23/2006
10430395  53736422       130         86   11/18/2006

[10430396 rows x 4 columns]

In [100]:
# convert systolic BPs >400 NaN
df_BP_impute['SYSTOLIC'][df_BP_impute['SYSTOLIC']>400] = np.nan
df_BP_impute.head()

/anaconda2/envs/bios8366/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


RUID  SYSTOLIC  DIASTOLIC Measure_date
0  50135262     150.0         80   01/09/2005
1  50135262     137.0         77   02/08/2007
2  50135262     137.0         78   02/08/2007
3  50135262     190.0         78   02/08/2007
4  50135262     190.0         77   02/08/2007

In [105]:
pd.set_option('display.max_rows', 100)
df_BP_impute.sort_values(by='SYSTOLIC',ascending=True).tail(100)

RUID  SYSTOLIC  DIASTOLIC Measure_date
9634703   53735891       NaN         56   11/26/2005
9705983   53735941       NaN         44   06/24/2013
9706025   53735941       NaN         68   06/24/2013
9706034   53735941       NaN         47   06/24/2013
9706057   53735941       NaN         57   06/24/2013
9706066   53735941       NaN         65   06/24/2013
9706077   53735941       NaN         51   06/24/2013
9706084   53735941       NaN         66   06/24/2013
9706126   53735941       NaN         56   06/24/2013
9706138   53735941       NaN         43   06/24/2013
9706172   53735941       NaN         80   06/24/2013
9706207   53735941       NaN         50   06/24/2013
9706224   53735941       NaN         60   06/24/2013
9706229   53735941       NaN         52   06/24/2013
9706238   53735941       NaN         54   06/24/2013
9706343   53735941       NaN         61   06/24/2013
9706349   53735941       NaN         67   06/24/2013
9706352   53735941       NaN         77   06/24/2013
9706357   53735941       NaN         62   06/24/2013
9733317   53735958       NaN         64   11/25/2012
9733319   53735958       NaN         53   11/25/2012
9733322   53735958       NaN         52   11/25/2012
9733340   53735958       NaN         56   11/25/2012
9733342   53735958       NaN         58   11/25/2012
9733362   53735958       NaN         50   11/25/2012
9733364   53735958       NaN         59   11/25/2012
9733399   53735958       NaN         43   11/25/2012
9733406   53735958       NaN         63   11/25/2012
9733412   53735958       NaN         72   11/25/2012
9735333   53735958       NaN         69   12/14/2012
9735337   53735958       NaN         66   12/14/2012
9735341   53735958       NaN         72   12/14/2012
9735353   53735958       NaN         60   12/14/2012
9735356   53735958       NaN         94   12/14/2012
9735364   53735958       NaN         71   12/14/2012
9735366   53735958       NaN          2   12/14/2012
9735377   53735958       NaN         62   12/14/2012
9853531   53736042       NaN         57   04/18/2013
9853544   53736042       NaN         58   04/18/2013
9853548   53736042       NaN         43   04/18/2013
9853557   53736042       NaN         51   04/18/2013
9853570   53736042       NaN         47   04/18/2013
9853571   53736042       NaN         55   04/18/2013
9853582   53736042       NaN         48   04/18/2013
9881184   53736058       NaN         53   02/28/2006
9881186   53736058       NaN          6   02/28/2006
9881188   53736058       NaN         55   02/28/2006
9901500   53736073       NaN         39   12/24/2012
9901502   53736073       NaN         50   12/24/2012
9901529   53736073       NaN         65   12/24/2012
9901538   53736073       NaN         52   12/24/2012
9901543   53736073       NaN         54   12/24/2012
9901544   53736073       NaN         48   12/24/2012
9901549   53736073       NaN         46   12/24/2012
9972652   53736120       NaN         56   09/10/2005
9972664   53736120       NaN         59   09/10/2005
9972670   53736120       NaN          9   09/10/2005
9972674   53736120       NaN         62   09/10/2005
9972675   53736120       NaN         64   09/10/2005
9972677   53736120       NaN         58   09/10/2005
9972680   53736120       NaN         70   09/10/2005
10019411  53736174       NaN          5   07/18/2008
10019412  53736174       NaN         43   07/18/2008
10019420  53736174       NaN         63   07/18/2008
10019421  53736174       NaN         56   07/18/2008
10019427  53736174       NaN         44   07/18/2008
10019428  53736174       NaN         52   07/18/2008
10036906  53736193       NaN         75   03/13/2009
10036911  53736193       NaN         81   03/13/2009
10036912  53736193       NaN         80   03/13/2009
10036952  53736193       NaN         77   03/13/2009
10036956  53736193       NaN         89   03/13/2009
10036970  53736193       NaN         82   03/13/2009
10036980  53736193       NaN         59   03/13/2009
10036983  53736193       NaN         69

## BMI

1) carry forward impute BMI for NaN values

In [107]:
df_BMI = pd.read_csv('./Data/FONNESBECK_BMI_20151202.csv',sep=',')
df_BMI.head()

RUID    BMI    Date_BMI  BMI_Weight  BMI_Height  Pregnancy_Indicator
0  50135262  41.43  01/09/2005      119.98      170.18                    0
1  50135262  22.86  02/11/2011       70.00      175.00                    0
2  50135262  43.07  02/12/2011      131.90      175.00                    0
3  50135262  43.06  02/12/2011      131.86      175.00                    0
4  50135262  41.47  02/12/2011      127.00      175.00                    0

In [109]:
pd.set_option('display.max_rows', 25)
df_BMI[df_BMI['BMI'].isna()]

RUID  BMI    Date_BMI  BMI_Weight  BMI_Height  Pregnancy_Indicator
744     50135624  NaN  01/21/2013         NaN      177.80                    0
1452    50136045  NaN  07/04/2014         NaN      165.10                    0
1676    50136072  NaN  12/24/2013         NaN      162.56                    0
3108    50138770  NaN  03/14/2013         NaN      167.64                    0
4164    50142794  NaN  12/19/2014         NaN      177.80                    0
5846    53719335  NaN  09/15/2012         NaN      182.88                    0
6902    53726156  NaN  02/21/2013         NaN      154.00                    0
7665    53727849  NaN  07/28/2014         NaN      182.00                    0
8744    53727891  NaN  08/28/2012         NaN      182.00                    0
10512   53727974  NaN  11/10/2013         NaN      118.00                    0
13195   53728087  NaN  03/09/2014         NaN      165.00                    0
13217   53728087  NaN  06/01/2012         NaN      160.02                    0
...          ...  ...         ...         ...         ...                  ...
255829  53736229  NaN  12/11/2012         NaN      173.99                    0
256169  53736238  NaN  05/12/2015         NaN       87.60                    0
256427  53736242  NaN  11/24/2013         NaN      187.96                    0
257036  53736272  NaN  06/01/2013         NaN      154.94                    0
257037  53736272  NaN  06/01/2013         NaN      154.90                    0
258047  53736290  NaN  10/16/2014         NaN      172.00                    0
258385  53736303  NaN  05/16/2013         NaN      154.94                    0
259130  53736313  NaN  03/07/2013         NaN      175.26                    0
260475  53736350  NaN  02/06/2014         NaN      170.18                    0
260837  53736366  NaN  06/13/2013         NaN      182.88                    0
260929  53736368  NaN  08/18/2012         NaN       97.00                    0
261386  53736379  NaN  02/05/2014         NaN      165.10                    0

[307 rows x 6 columns]

In [113]:
pd.set_option('display.max_rows', 50)
df_BMI[df_BMI['RUID']==50135361].head(50)

RUID     BMI    Date_BMI  BMI_Weight  BMI_Height  Pregnancy_Indicator
71   50135361   27.89  01/07/2011       90.72      180.34                    0
72   50135361   27.71  01/09/2012       97.90      187.96                    0
73   50135361   27.69  01/10/2012       97.84      187.96                    0
74   50135361   27.68  01/10/2012       97.80      187.96                    0
75   50135361   31.53  01/15/2007      111.40      187.96                    0
76   50135361   31.50  01/19/2007      111.27      187.96                    0
77   50135361   33.58  01/22/2005      115.44      185.42                    0
78   50135361   34.90  01/25/2003      116.71      182.88                    0
79   50135361   33.81  01/28/2005      116.23      185.42                    0
80   50135361   30.13  01/29/2009      103.60      185.42                    0
81   50135361   37.16  02/01/2002      124.28      182.88                    0
82   50135361   27.70  02/04/2010       95.25      185.42                    0
83   50135361   29.71  02/04/2011       96.62      180.34                    0
84   50135361   36.68  02/17/2006      119.29      180.34                    0
85   50135361   28.55  02/18/2010       98.15      185.42                    0
86   50135361   31.94  02/19/2009      103.87      180.34                    0
87   50135361   26.96  02/20/2012       95.25      187.96                    0
88   50135361   26.71  02/23/2012       94.35      187.96                    0
89   50135361   26.96  02/24/2012       95.25      187.96                    0
90   50135361   29.71  03/05/2011       96.62      180.34                    0
91   50135361   32.35  03/06/2009      105.20      180.34                    0
92   50135361   31.55  03/08/2009      102.60      180.34                    0
93   50135361   31.58  03/09/2009      102.70      180.34                    0
94   50135361   25.94  03/14/2010       91.63      187.96                    0
95   50135361   34.44  03/17/2003      115.17      182.88                    0
96   50135361   33.38  03/18/2005      117.93      187.96                    0
97   50135361   33.67  03/19/2005      118.95      187.96                    0
98   50135361   36.73  03/19/2006      119.44      180.34                    0
99   50135361   33.56  03/20/2004      115.38      185.42                    0
100  50135361   34.84  03/22/2003      116.53      182.88                    0
101  50135361   32.27  03/29/2007      113.99      187.96                    0
102  50135361   31.67  03/29/2009      103.00      180.34                    0
103  50135361   32.78  04/02/2009      106.59      180.34                    0
104  50135361   27.22  04/05/2012       96.16      187.96                    0
105  50135361   27.44  05/01/2010       94.35      185.42                    0
106  50135361   27.45  05/02/2010       94.37      185.42                    0
107  50135361   34.85  05/09/2003      116.57      182.88                    0
108  50135361  390.59  05/10/2007      111.13       53.34                    0
109  50135361   29.09  05/11/2009      100.02      185.42                    0
110  50135361   33.63  05/13/2004      115.62      185.42                    0
111  50135361   34.70  05/13/2006      119.29      185.42                    0
112  50135361  402.54  05/21/2007      114.53       53.34                    0
113  50135361   36.15  05/24/2002      120.91      182.88                    0
114  50135361   26.82  05/24/2009       92.22      185.42                    0
115  50135361   27.52  05/25/2009       94.62      185.42                    0
116  50135361   27.14  05/26/2009       93.30      185.42                    0
117  50135361   34.44  05/27/2006      118.39      185.42                    0
118  50135361   32.18  06/09/2012       96.00      172.72                    0
119  50135361   32.19  06/09/2012       96.00      172.70                    0
120  50135361   32.93  06/10/2012       98.20      172.70  

there are non-physiological values of BMI for patients who have otherwise physiological BMIs, so would have to impute these values as well

2) explore patients with pregnancy_indicator = 1

In [120]:
df_BMI['RUID'].value_counts()

53728553    2718
53728953    1589
53730165    1450
53734564    1391
53729836    1370
53729875    1277
53731544     914
53735080     875
53728324     755
53735673     744
53731053     686
53735253     681
53735412     667
53733216     664
53734017     606
53733189     534
53731781     415
53732315     391
53736180     390
53732346     369
53732544     367
53734444     366
53730229     349
53736315     347
53730766     340
            ... 
53731373       1
53730341       1
53732898       1
53731619       1
53734946       1
53732342       1
53729050       1
53730763       1
53728289       1
53735569       1
53729056       1
53729906       1
53729636       1
53729569       1
53727845       1
53729268       1
53734948       1
53728101       1
53734442       1
53730998       1
53734067       1
53732911       1
53736110       1
53730906       1
53736124       1
Name: RUID, Length: 6634, dtype: int64

In [117]:
df_BMI_preg = df_BMI[df_BMI['Pregnancy_Indicator']==1]

In [118]:
df_BMI_preg['RUID'].nunique()

608

there are 608/6634 patients with pregnancy_indicator = 1, but most likely these patients are pregnant all of the time, i.e. may be able to use the data 

to do: need to impute NaN and non-physiological values of BMI with carry forward

leave pregnancy_indicator as its own variable, and random forest would be expected to take care of interaction between variables

## EGFR

EGFR: histogram is what we expect during last meeting and is trajectory that they observe when people head towards dialysis to see bump at the end

to-do:  
* may want to use last value in model for various measures
* but this may change to look at trends, i.e. look at variance
* create a helper function that starts dumps out dataframe with 1. last value, 2. median, 3. standard deviation, 4. IQR, 5. date range